packages/router-core/src/path.ts

```diff
@@ -796,33 +805,111 @@ function isMatch(
+                  : futureRouteSegment.value.toLowerCase() ===
+                    baseSegment.value.toLowerCase()
+              ) {
+                // so we should skip this optional parameter
+                shouldMatchOptional = false
+              }
+              break
+            }
+
+            // where consecutive optional params are used, we can break early.
+            // preference is given to the first optional param
+            if (futureRouteSegment?.type === SEGMENT_TYPE_OPTIONAL_PARAM) {
+              if (
+                remainingRequiredRouteSegmentCount >= remainingBaseSegmentCount
+              ) {
+                shouldMatchOptional = false
+              }
               break
             }
 
-            // If we encounter a required param or wildcard, stop looking ahead
+            // this if is not required for the rest of the logic, but it's useful to know what type of future segment we're looking at'
             if (
               futureRouteSegment?.type === SEGMENT_TYPE_PARAM ||
               futureRouteSegment?.type === SEGMENT_TYPE_WILDCARD
             ) {
-              if (baseSegments.length < routeSegments.length) {
-                shouldMatchOptional = false
+              const followingRouteSegment = routeSegments[lookAhead + 1]
+
+              let isMatchedFurtherDown = false
+
+              // since we know there are remaining optionals, we look to the segment following the next.
+              // if further segments are required paths, then we can possibly match further optionals based on the url pattern. to do this, we match the remaining paths
+              // if all that follows are wildcards/required params and/or optionals params we continue matching on a first-case basis, and optionals further down are unmatched.
+              if (
+                remainingOptionals &&
+                followingRouteSegment &&
+                (followingRouteSegment.type === SEGMENT_TYPE_PATHNAME ||
+                  (followingRouteSegment.type === SEGMENT_TYPE_OPTIONAL_PARAM &&
+                    (followingRouteSegment.prefixSegment ||
+                      followingRouteSegment.suffixSegment)))
+              ) {
+                const remainingRouteSegments = routeSegments.slice(
+                  lookAhead + 1,
+                )
+
+                const remainingRouteSegmentLength =
+                  remainingRouteSegments.slice(-1)[0]?.value === '/'
+                    ? remainingRouteSegments.length - 1
+                    : remainingRouteSegments.length
+
+                const remainingBaseSegments = baseSegments.slice(baseIndex + 1)
+
+                isMatchedFurtherDown =
+                  remainingRouteSegmentLength ===
+                    remainingBaseSegments.length &&
+                  isMatch(
+                    remainingBaseSegments,
+                    remainingRouteSegments,
+                    { ...params },
```

Comment:
I don't think it needs to inherit params
```suggestion
                    {},
```

Reply:
params gets modified in isMatch when the match is found, for this here we only want to confirm a match exists and don't want it to mutate the params, hence the spread.

Reply:
but it's never reading from it either right? So if we're not gonna use it, and it doesn't need it, why spread at all? Why not just pass in an empty object?